File: web/src/lib/readiness.ts
/**
 * Per-site readiness — the gating engine for a network site, driven by **domain activation**
 * (#1220 / #1223).
 *
 * The network is *additive*: a site is defined by the **domains that actually have a story
 * there**, not by its deficits against Lima's full section taxonomy. Readiness is computed **in
 * Python at export** (`bosc.site.readiness`) and written into the bundle `manifest.json` as a
 * `readiness` block (five domains × `absent|seeded|live`, plus a derived `tier`). This module reads
 * that block — it is the SSOT for the site-level gating and the chrome tier, and there is no
 * "reference site ⇒ everything available" shortcut: Lima renders as available because its manifest
 * says every domain is `live`, not because it is special-cased.
 *
 * Sections gate in two bands:
 *   - **Primary sections** (`record` / `places` / `environment` / `economy`) read their parent
 *     **domain** state from the manifest block — a domain with any evidence (`seeded` or `live`)
 *     opens them. This is what lets a Backdrop-tier peer (floor data only) render a real environment +
 *     economy page instead of a wall of locks.
 *   - **Leaf facets** (`timeline` / `people` / `exhibits` / `story` / `leads`) additionally require
 *     their own feed/registry signal, so an active domain never opens an *empty* facet page (a
 *     timeline with no events reads as a lock + needs-board ask, not a barren page).
 * Two sections are network-global rather than per-site: `reports` (the `docs/` long-form the
 * reference build hosts) stays reference-only (`isReferenceSite`, the surviving network-global-host
 * role, #1220); `environment` also locks when the facility's cooling method is undisclosed (#1057).
 */
import { hasFeed, loadFeed, loadManifest } from "./bundle";
import type { DomainState, Readiness, SiteTier } from "./bundle";
import type { ScenarioResult } from "./feeds";
import { LIMA_SLUG } from "./routes";
import { surfacedStories } from "./sites";

export type { DomainState, SiteTier } from "./bundle";

/** The five activation domains (`bosc.site.readiness.Domain`). */
export type Domain = "backdrop" | "facility" | "places" | "record" | "story";

/**
 * The live reference build (Lima) hosts the network-global content — the `docs/` narrative that
 * the `reports` section reads, the cross-site hypothesis matrix, the whole-data-tier catalog. This
 * is the **network-global-host role** (the peer of `bosc.sites.is_reference_site`), NOT a readiness
 * backdoor: it no longer forces every section available (#1220). Only `reports` still keys off it.
 */
export function isReferenceSite(slug: string): boolean {
  return slug === LIMA_SLUG;
}

export type SectionStatus = "available" | "locked";

/** A gateable site section — a top-level destination under `/network/<site>`. */
export type ReadinessSection =
  | "record"
  | "timeline"
  | "people"
  | "places"
  | "exhibits"
  | "environment"
  | "economy"
  | "reports"
  | "story"
  | "leads"
  | "contacts";

/** Display + lock copy for each section. `holds` answers "what lands here once we have sources"
 *  — shown on the lock so an empty section reads as *awaiting a source*, not broken or barren. */
export const SECTION_META: Record<ReadinessSection, { label: string; holds: string }> = {
  record: {
    label: "The record",
    holds: "the source documents, structured records, and the entities drawn from them",
  },
  timeline: {
    label: "Timeline",
    holds: "the dated events reconstructed from the record — permits, filings, meetings",
  },
  people: {
    label: "People",
    holds: "the actors named in the record, each tied back to the documents that name them",
  },
  places: {
    label: "Places",
    holds: "the parcels, facilities, and waters the record locates",
  },
  exhibits: {
    label: "Exhibits",
    holds: "the curated source exhibits — the documents that carry the keystone figures",
  },
  environment: {
    label: "The environment",
    holds:
      "the hydrology, imagery, air-dispersion, and toxics-release picture for the receiving water and air",
  },
  economy: {
    label: "Economy",
    holds: "the load, grid, and economic-baseline reads for the site",
  },
  reports: {
    label: "Reports",
    holds: "the long-form analysis built over the corpus, once the corpus supports it",
  },
  story: {
    label: "The story",
    holds: "the guided walk that teaches the record one document at a time",
  },
  leads: {
    label: "Open leads",
    holds: "the open questions and the source data we're seeking for this site",
  },
  contacts: {
    label: "Contacts",
    holds:
      "the people and bodies to reach on this site — the officials, organizers, and groups the record names",
  },
};

/** All-absent readiness — the safe fallback when a bundle predates the `readiness` block (contract
 *  < 1.17.0) or a synthetic fixture omits it: sections lock (degrade), nothing crashes. */
const ABSENT_READINESS: Readiness = {
  tier: "stub",
  domains: { backdrop: "absent", facility: "absent", places: "absent", record: "absent", story: "absent" },
};

/** The site's computed readiness block, read straight from its bundle manifest (#1220). */
export function siteReadinessBlock(slug: string): Readiness {
  return loadManifest(slug).readiness ?? ABSENT_READINESS;
}

/** The five domains' `absent|seeded|live` states for a site (the manifest block, domain axis). */
export function siteDomainStates(slug: string): Record<Domain, DomainState> {
  return siteReadinessBlock(slug).domains;
}

/** A site's readiness tier (`stub|backdrop|case|reference`) — the chrome/tab tier reads this. */
export function siteTier(slug: string): SiteTier {
  return siteReadinessBlock(slug).tier;
}

/** Count a feed's rows from the manifest (0 when the feed is absent). */
function feedCount(slug: string, name: string): number {
  return loadManifest(slug).feeds.find((f) => f.name === name)?.count ?? 0;
}

/**
 * Whether the site's water math rests on an **undisclosed cooling method** (#1057).
 *
 * A `hydrology-scenarios` row whose `cooling_model` is `unknown` (equivalently, whose basis
 * says `method_disclosed: false`) carries a bracketed range across candidate archetypes, not
 * an estimate — the facility exists but no record says how it rejects heat. Rendering its
 * single consumptive/7Q10-multiple headline as if confirmed would fabricate the site's most
 * load-bearing number, so the environment section locks and the needs board asks for the
 * disclosure instead. Content-based (reads feed rows), so it stands apart from the domain block.
 */
export function coolingMethodUndisclosed(slug: string): boolean {
  if (!hasFeed("hydrology-scenarios", slug)) return false;
  const rows = loadFeed<ScenarioResult[]>("hydrology-scenarios", slug);
  return rows.some(
    (r) =>
      (r.cooling_model ?? r.scenario.cooling_model) === "unknown" ||
      r.scenario.basis?.method_disclosed === false,
  );
}

/** A domain carries evidence (is worth opening its sections for) when it is `seeded` or `live`. */
function domainPresent(slug: string, domain: Domain): boolean {
  return siteDomainStates(slug)[domain] !== "absent";
}

/**
 * Whether a section has enough of *this site's* own data to stand on its own — the additive gate.
 * Primary sections read the manifest `readiness` block (their parent domain); leaf facets add a
 * feed/registry check so an active domain never opens an empty page. Deterministic per bundle.
 */
function hasEnough(section: ReadinessSection, slug: string): boolean {
  switch (section) {
    // --- primary sections: gated by their parent domain's activation state (the manifest block) ---
    case "record":
      return domainPresent(slug, "record");
    case "places":
      return domainPresent(slug, "places");
    case "economy":
      return domainPresent(slug, "backdrop");
    case "environment":
      // The floor's environment read (hydrology + toxics) opens with the backdrop domain — but an
      // undisclosed cooling method locks it outright (#1057): its scenario rows are bracketed
      // ranges, and no fabricated single-figure headline may stand in.
      return domainPresent(slug, "backdrop") && !coolingMethodUndisclosed(slug);
    // --- leaf facets: the domain plus the facet's own feed/registry signal (no empty pages) ---
    case "timeline":
      return domainPresent(slug, "record") && feedCount(slug, "timeline") > 0;
    case "people":
      return domainPresent(slug, "record") && feedCount(slug, "people") > 0;
    case "exhibits":
      return domainPresent(slug, "record") && feedCount(slug, "exhibits") > 0;
    case "story":
      // The guided walk needs a *surfaced* story (registered in the `sites.ts` overlay and not
      // hidden) — a leads-only story domain (Urbana) has no walk to open, and Lima's hidden Project
      // BOSC walk (#1256) locks here too, so its story tab/hub CTA drop in the early build state.
      return surfacedStories(slug).length > 0;
    case "leads":
      // The leads board is feed-driven per site (#796); the reference build also hosts the
      // network-global curated board.
      return feedCount(slug, "leads") > 0 || isReferenceSite(slug);
    case "contacts":
      // The contacts directory is purely feed-driven: a site opens it only once it ships its own
      // committed `contacts` feed — never borrowing another site's contacts.
      return feedCount(slug, "contacts") > 0;
    // --- network-global: the reference build hosts the long-form `docs/` narrative ---
    case "reports":
      return isReferenceSite(slug);
  }
}

/** A section's status for a site: `available` (render its real data) or `locked` (show the lock). */
export function sectionStatus(slug: string, section: ReadinessSection): SectionStatus {
  return hasEnough(section, slug) ? "available" : "locked";
}

/** Convenience: is this section ready to render for the site? */
export function isAvailable(slug: string, section: ReadinessSection): boolean {
  return sectionStatus(slug, section) === "available";
}

/** Every section's status for a site — the full readiness map (the model the pages + nav read). */
export function siteReadiness(slug: string): Record<ReadinessSection, SectionStatus> {
  const out = {} as Record<ReadinessSection, SectionStatus>;
  for (const section of Object.keys(SECTION_META) as ReadinessSection[]) {
    out[section] = sectionStatus(slug, section);
  }
  return out;
}

/** The sections currently locked for a site (empty for a site whose domains are all lit). */
export function lockedSections(slug: string): ReadinessSection[] {
  return (Object.keys(SECTION_META) as ReadinessSection[]).filter((s) => sectionStatus(slug, s) === "locked");
}

File: src/watermark/site/readiness.py
"""Site readiness by **domain activation** — the SSOT the frontend reads (#1220 / #1221).

The BOSC network is *additive*, not *subtractive*: a site is defined by the **domains that
actually have a story there**, not by its deficits against Lima's full ~10-section taxonomy.
This module classifies a site into five domains, each in one of three states
(``absent | seeded | live``), from ``(SiteProfile, feed counts)`` alone — no hardcoded
Lima/Allen values — and derives a single site **tier** from that vector.

It is the Python peer of ``web/packages/core/src/readiness.ts``. Here it is import-only; #1222 wires it
into ``watermark export`` so the computed block is written into ``manifest.json`` and the
frontend becomes a thin reader (#1223). Because the readiness is recomputed at every export,
it is a **standing** property — it goes up when a source lands and down when one dries up,
without re-running ``onboard``.

The domain model (the two problems #1220 fixes — see the epic):

============  ================================================  =========================
domain        live when                                         signal
============  ================================================  =========================
**backdrop**  every always-pull floor feed present              floor feeds (FIPS/state/grid)
**facility**  ``SiteProfile.facility`` instrument-grounded      profile facility grounding
**places**    committed campus **or** enclave geometry          ``geo/campus``/``geo/enclave``
**record**    extracted corpus over threshold                   records+documents+entities
**inquiry**   the study ANSWERS, over a record that exists      ``impact-study`` verdicts
============  ================================================  =========================

``inquiry`` was called ``story`` until #1971 (epic #1968), and the rename is the point rather than
cosmetic. Its predicate was ``slug in STORY_SLUGS and leads`` — a hand-maintained Python mirror of a
TypeScript overlay of MDX directories. **It was the only domain whose signal was "did a human author
prose."** It measured editorial output, it gated the tier, and so it made a walk the price of a
site's fifth domain: van-wert carried three merged investigations and read ``absent``, while findlay
read ``live`` for a walk that was ``comingSoon`` and could not be opened.

It now measures whether the site's own study **answers**, from the ``impact-study`` verdicts that
same export just computed — data the way every other domain is data. ``STORY_SLUGS`` is deleted;
the leads feed keeps its own facet signal and is no longer ANDed in.

Two conditions, and the second is what makes it about the record:

* a substantive count — chapters reading ``data`` or ``partial``, and
* **record-backed** — at least one of the two corpus-keyed chapters (``assembly`` / ``governance``,
  #1969) is itself substantive.

The second exists because the count alone is not a record signal. Nine of the fifteen chapters
derive from connector pulls, the facility profile and the grid backdrop, so springfield (zero
records) out-scored several worked corpora on count alone before #1969 landed the two chapters that
read the corpus. Requiring one of those to be substantive is what keeps a floor-only pull from
reading as an answered study.

**It no longer gates the tier** (see :func:`site_tier`).

Facility is graded on **documentary depth** (#1630), not on one economics feed: a facility whose
load is grounded in a primary instrument (an air permit, or a filed disclosure) or whose cooling is
document-disclosed is ``live``; a facility on the record only by name / announcement / site-plan
SCREENING ([inference]) is ``seeded``. The old rule keyed ``live`` on the ``economics-demand-pressure``
feed, collapsing permit-grounded (Lima) and screening-only (Urbana) evidence onto one label.

The trigger *is* the evidence: nothing above the floor scaffolds. A partial site **locks**
the domains it lacks and asks for the source, per the repo's spine — never fabricate a value
to make a thin peer look complete.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Literal, get_args

from watermark.sites import SiteProfile

# --- Vocabulary -----------------------------------------------------------------------------
State = Literal["absent", "seeded", "live"]
Domain = Literal["backdrop", "facility", "places", "record", "inquiry"]
Tier = Literal["stub", "backdrop", "case", "reference"]

DOMAINS: tuple[Domain, ...] = get_args(Domain)

# --- Activation signals (named constants so #1223 can mirror them exactly) -------------------
# The always-pull floor: the connectors keyed only by coordinates / county FIPS / state / the
# site's own EIA-861 utility number, so they carry zero curation and no fabrication risk.
# NASA-POWER climatology / Atlas-14 / WBD / SSURGO feed hydrology but emit no distinct floor
# feed of their own (``hydrology-scenarios`` is facility-gated), so RSEI stands as a FIPS-keyed
# floor signal alongside the two economic ones.
#
# ``grid`` joined the floor in #1642 (GP-E E1). The grid backdrop — *whose* utility serves the
# site, within which balancing authority, and how big each is — is a **property of the place**,
# not of any campus proposed on it: ``derive_grid_profile`` emits it with ``load_share=None`` for
# a facility-less site precisely so a thin peer still gets the real electric-service chain. It is
# the same always-pull shape as its neighbours here (state/utility-keyed connectors, no curation),
# and until now it was written only to a CLI reference file and never reached the site at all,
# which is why the frontend hardcoded Lima's denominators.
#
# These are object feeds, which serialize with ``count == 1`` when present — so ``_count(...) > 0``
# tests *presence*, and presence must mean *content*. The exporter drops a genuinely-empty
# inventory (``watermark.site.export``: an RSEI feed with zero facilities, an economic baseline
# with no sectors, a grid profile with zeroed utility/BA denominators) so it is absent, not a
# ``count == 1`` shell that floats backdrop to ``live`` on an empty inventory (#1364).
ECONOMICS_BASELINE_FEED = "economics-baseline"
CONSUMER_ENERGY_FEED = "consumer-energy"
RSEI_FEED = "rsei"
GRID_FEED = "grid"
BACKDROP_FLOOR_FEEDS: tuple[str, ...] = (
    ECONOMICS_BASELINE_FEED,
    CONSUMER_ENERGY_FEED,
    RSEI_FEED,
    GRID_FEED,
)

# The facility demand→price-pressure feed, present only where a facility is disclosed (#1220:
# "demand-pressure missing across 22 sites is not a gap — it's facility-gated"). Since #1630 this
# is the facility domain's **leaf feed** — the facility-facing data the frontend renders — NOT the
# Python live-trigger: ``_facility_state`` now grades on the profile's own facility evidence, and
# the frontend's thin reader gates the demand-pressure section on ``domainPresent(…, "facility")``
# plus ``hasFeed(FACILITY_FEED)`` so an active facility never opens an empty page. Kept in
# ``READINESS_FEED_NAMES`` so the coupling guard still catches a rename that would break that gate.
FACILITY_FEED = "economics-demand-pressure"

# Committed campus/footprint geometry (the exported parcel/footprint layer) vs. place *records*
# without geometry. ``gis_parcel`` is deliberately NOT the seeded signal: the statewide Ohio
# schema is set on nearly every profile (stubs included), so it does not discriminate — only
# actually-committed geometry (which surfaces as ``geo/campus``) or place records do.
PLACES_GEOMETRY_FEED = "geo/campus"
# The **non-CAMA** land path (#1664). ``places`` gated on ``geo/campus`` alone made the domain
# structurally unreachable for a federal enclave: a base is off the county tax rolls, so no
# county parcel layer will ever carry it and no amount of research would ever produce a parcel
# assemblage. That was a gap in the model, not a gap in the evidence — the DoD MIRTA site
# register carries the boundary, and committed register geometry activates the domain on exactly
# the same terms as committed parcel geometry: the trigger is still the evidence, and still
# geometry a map can actually be drawn from.
PLACES_ENCLAVE_FEED = "geo/enclave"
PLACES_GEOMETRY_FEEDS: tuple[str, ...] = (PLACES_GEOMETRY_FEED, PLACES_ENCLAVE_FEED)
PLACES_RECORD_FEED = "places"

# The per-site extracted **document corpus** — the genuinely site-scoped record feeds. Only
# ``records`` (extracted RecordItems) and ``documents`` (per-site DocumentCollectionItems) are
# corpus-scoped and clean. Deliberately EXCLUDED: ``entities`` (the entity feed merges the
# network-global defense-contractor cross-match — e.g. "General Dynamics" surfaces on every site —
# so it does not discriminate a real corpus), ``concepts`` and ``defense-contractors`` (both
# network-global boilerplate present on every bundle). A site with RSEI-derived facility entities
# but no document corpus is correctly a Backdrop site, not a record case.
#
# Only ``records`` (validated ``RecordItem``s — actual extracted content) can float the domain to
# ``live``. ``documents`` counts raw source scans by first-level COLLECTION dir (one row per
# collection, not per file — ``watermark.site.documents``), so it measures containers, not content:
# a thin site with un-extracted scans in ≥2 collection dirs would otherwise sum to ``live`` with
# zero extractions (#1364). Catalogued-but-unworked scans SEED the record domain; they never lift
# it — "let it lock and ask for the source" rather than read present-but-empty as complete.
RECORD_LIVE_FEED = "records"
DOCUMENTS_FEED = "documents"
RECORD_SEED_FEEDS: tuple[str, ...] = (RECORD_LIVE_FEED, DOCUMENTS_FEED)
# ``records`` at/above this is ``live``; any record signal below it is ``seeded``; nothing is
# ``absent``. Two extracted items is the "real worked corpus" bar (a lone stray extraction seeds).
RECORD_LIVE_THRESHOLD = 2

# The leads feed — the open-questions / source-solicitation board for the site. It keeps its own
# facet signal in the frontend and is deliberately NOT part of any domain predicate (#1971): ANDing
# it into the old ``story`` domain is what let a site reach a higher tier by committing a YAML file.
LEADS_FEED = "leads"

# --- The ``inquiry`` signal (#1971) ----------------------------------------------------------
# Study chapter verdicts that count as the study having said something. ``gap`` is content — the
# study renders the absence AS a finding and never locks — but it is not an answer, and ``na`` is
# the watch state for a site with no disclosed project.
SUBSTANTIVE_STATUSES: frozenset[str] = frozenset({"data", "partial"})
# The two corpus-keyed chapters (#1969) — the only ones whose verdict is screened against the
# site's OWN extracted records rather than a connector pull or the facility profile.
RECORD_KEYED_CHAPTERS: tuple[str, ...] = ("assembly", "governance")
# Substantive chapters needed for ``live``. Eight of fifteen is the bar that separates a worked
# corpus from a thin one on the committed cohort (van-wert 9, ottawa 7) — deliberately a plain
# count, because a RATIO would flatter a facility-less site: its project-dependent chapters read
# ``na`` and drop out of the denominator, so the hardest chapters would stop counting against it.
INQUIRY_LIVE_THRESHOLD = 8
# Substantive chapters needed for ``seeded`` without any record-keyed chapter — a floor-only site
# whose study still reports a real labor baseline and grid backdrop has said *something*.
INQUIRY_SEED_THRESHOLD = 5

# Every manifest feed name this module keys a domain on — the single enumerable coupling to the
# exporter's feed spec (``watermark.site.export``). ``export.py`` shares these very constants for
# the feeds it can name directly; the composed geo feed (``geo/campus``) can't, so this set is what
# ``tests/test_site_bundle.py`` asserts the exporter still produces — a rename in ``export.py``
# that skips ``readiness.py`` then drops a feed out of the produced set and fails the guard, rather
# than silently dropping every site's facility/backdrop to ``seeded`` with green tests (#1631).
READINESS_FEED_NAMES: frozenset[str] = frozenset(
    {
        *BACKDROP_FLOOR_FEEDS,
        FACILITY_FEED,
        *PLACES_GEOMETRY_FEEDS,
        PLACES_RECORD_FEED,
        *RECORD_SEED_FEEDS,
        LEADS_FEED,
    }
)


def _count(feed_counts: Mapping[str, int], name: str) -> int:
    """A feed's row count, 0 when absent."""
    return feed_counts.get(name, 0)


def _tri(present: int, total: int) -> State:
    """``live`` when all ``total`` signals present, ``seeded`` when some, ``absent`` when none."""
    if present >= total:
        return "live"
    return "seeded" if present > 0 else "absent"


def _backdrop_state(feed_counts: Mapping[str, int]) -> State:
    present = sum(1 for f in BACKDROP_FLOOR_FEEDS if _count(feed_counts, f) > 0)
    return _tri(present, len(BACKDROP_FLOOR_FEEDS))


def _facility_state(profile: SiteProfile) -> State:
    # No disclosed facility → absent (grid backdrop only, no fabricated campus load share).
    fac = profile.facility
    if fac is None:
        return "absent"
    # Documentary depth grades live vs seeded (#1630): an instrument-grounded facility — an air
    # permit / filed disclosure grounding the load, or a document-disclosed cooling mechanism —
    # reads `live`; a facility on the record only by name / announcement / site-plan SCREENING
    # ([inference]) reads `seeded`. Keyed on the profile's OWN facility evidence, not on the
    # presence of the demand-pressure feed (which the pre-#1630 rule collapsed permit and screening
    # onto — Urbana's floor-area [inference] and Lima's permit-grounded load both read `live`).
    return "live" if fac.is_instrument_grounded else "seeded"


def _places_state(feed_counts: Mapping[str, int]) -> State:
    # Committed campus parcels OR a committed federal-enclave boundary (#1664) — either is
    # drawable geometry for the place the site is about.
    if any(_count(feed_counts, f) > 0 for f in PLACES_GEOMETRY_FEEDS):
        return "live"
    # Place *records* without committed footprint geometry — some spatial signal, not enough
    # to draw the map with.
    return "seeded" if _count(feed_counts, PLACES_RECORD_FEED) > 0 else "absent"


def _record_state(feed_counts: Mapping[str, int]) -> State:
    # Live is gated on extracted content (``records``) alone; catalogued source scans
    # (``documents``, counted by collection dir) only ever seed the domain (#1364).
    if _count(feed_counts, RECORD_LIVE_FEED) >= RECORD_LIVE_THRESHOLD:
        return "live"
    seeded = any(_count(feed_counts, f) > 0 for f in RECORD_SEED_FEEDS)
    return "seeded" if seeded else "absent"


def _inquiry_state(study_statuses: Mapping[str, str] | None) -> State:
    """Whether this site's own study **answers** — ``impact-study`` verdicts, not authored prose.

    ``study_statuses`` maps a chapter id to its verdict. ``None`` means the caller had no study to
    read (a bundle predating the feed, or a synthetic fixture), which degrades to ``absent`` rather
    than guessing — the same "let it lock and ask for the source" rule the rest of this module runs.
    """
    if not study_statuses:
        return "absent"
    substantive = sum(1 for s in study_statuses.values() if s in SUBSTANTIVE_STATUSES)
    # Record-backed: the study says something the site's OWN extracted corpus grounds. Without
    # this, a site with zero records rides connector-derived chapters to a respectable count.
    record_backed = any(
        study_statuses.get(c) in SUBSTANTIVE_STATUSES for c in RECORD_KEYED_CHAPTERS
    )
    if record_backed and substantive >= INQUIRY_LIVE_THRESHOLD:
        return "live"
    if record_backed or substantive >= INQUIRY_SEED_THRESHOLD:
        return "seeded"
    return "absent"


def domain_states(
    profile: SiteProfile,
    feed_counts: Mapping[str, int],
    study_statuses: Mapping[str, str] | None = None,
) -> dict[Domain, State]:
    """Each domain's ``absent|seeded|live`` state for a site, from its profile + feed counts.

    ``feed_counts`` maps a manifest feed name to its row count (as the manifest's ``feeds[]``
    carry). Pure and deterministic per ``(profile, counts, study_statuses)`` — the same inputs the
    frontend reads back out of ``manifest.readiness``.

    ``study_statuses`` is optional **on purpose, and the default is load-bearing.** The exporter
    builds the ``impact-study`` feed by calling back into this function for the FACILITY state (the
    `project` chapter's probe reads it), so a mandatory study argument would close a cycle. It does
    not actually close, because ``inquiry`` is the only domain that reads the study and ``facility``
    reads the profile alone — so the exporter calls this once without a study to build the feed, and
    once with it to write the manifest block.
    """
    return {
        "backdrop": _backdrop_state(feed_counts),
        "facility": _facility_state(profile),
        "places": _places_state(feed_counts),
        "record": _record_state(feed_counts),
        "inquiry": _inquiry_state(study_statuses),
    }


# The domains the TIER is derived from (#1971). ``inquiry`` is reported beside them and never
# gates: the tier measures **the record a site has**, not how much of it has been written up. While
# the old ``story`` domain sat in this vector it was the terminal blocker on every site — so a tier
# moved when someone committed a leads YAML and registered an unreadable walk, and did NOT move for
# three merged investigations. A domain whose signal is derived FROM the other four (inquiry reads
# study verdicts, which read the feeds) would also double-count them if it gated.
TIER_DOMAINS: tuple[Domain, ...] = ("backdrop", "facility", "places", "record")


def site_tier(states: Mapping[Domain, State]) -> Tier:
    """Derive the site tier from its record-bearing domain vector (``TIER_DOMAINS``).

    * ``reference`` — every record-bearing domain live. This is the **readiness** tier and is
      distinct from ``is_reference_site`` (the network-global-host *role* — routed-hydrograph,
      hypothesis matrix, catalog, concepts), which is not a readiness backdoor.
    * ``case`` — the floor is live and at least one above-floor domain is live (Urbana:
      places+record).
    * ``backdrop`` — the floor is live but nothing above it is.
    * ``stub`` — not even the floor is live (profile only, ~zero data).

    ``inquiry`` is deliberately absent from every branch below. Dropping it promotes exactly two
    sites on the committed cohort — fort-wayne and wpafb, both already carrying all four
    record-bearing domains live — and demotes none.
    """
    if all(states[d] == "live" for d in TIER_DOMAINS):
        return "reference"
    if states["backdrop"] != "live":
        return "stub"
    above_floor = [d for d in TIER_DOMAINS if d != "backdrop"]
    if any(states[d] == "live" for d in above_floor):
        return "case"
    return "backdrop"


def compute_readiness(
    profile: SiteProfile,
    feed_counts: Mapping[str, int],
    study_statuses: Mapping[str, str] | None = None,
) -> dict[str, object]:
    """The full readiness block: per-domain states + the derived tier.

    The shape #1222 writes into ``manifest.json`` and #1223 reads: ``{"domains": {...},
    "tier": ...}``.
    """
    states = domain_states(profile, feed_counts, study_statuses)
    return {"domains": dict(states), "tier": site_tier(states)}
